### ISSUE TITLE
Feature request: calculation of duration and/or dtend/due

### ISSUE THREAD

**Issue by tobixen** [collective/icalendar#496]
According to the RFC, an icalendar VEVENT can have the property DURATION or the property DTEND, but it's not allowed to have both set.  I believe the intention is that an event with DTEND set one hour after the DTSTART is to be considered equivalent with an event with DURATION set to PT30M.

The same applies to tasks, they can have a DURATION or a DUE, but never both.  I suppose the intention there also is that a task with DTSTART and DURATION set should be considered equivalent with a task with DSTART and DUE set (though the tasks are a bit more muddy in my opinion - say there is three hours of paperwork that needs to be done before the end of the year, in my opinion it would make sense to set DUE to 2023-01-01, DURATION to three hours ... but one would probably want to start with it in early December, at least some days before the new years eve).

In the caldav library I now have methods `get_duration()` , `get_due()`, `set_due()` and `set_duration()`, which will attempt to do the right thing - like, `get_duration()` will yield DURATION if it's set, otherwise `DTEND-DTSTART`.  It makes sense generalize it or also add `get_dtend` and `set_dtend`.  Anyway ... this is logic that I need, but I feel it does not belong in the caldav library.  If anything, it belongs in the icalendar library - or perhaps a separate new package like icalendar_util.  (or perhaps there already exists some package like that?)

I will gladly contribute with a pull request on this one, but not without feedback that it's wanted and appropriate to put it into the icalendar library.

**Comment by niccokunzmann** [comment]
I like it. Let's make it happen. Deduplicate code, TDD. I have the same in recurring-ical-events. Please also note the PERIOD value type. It is good to get this straight for everyone.

If we can pass an rdate value, replacement start or end etc to this, that would be much appreciated later. I have PERIOD value type support still open.

It would be nice to use DUE or DTEND or PERIOD for the setter depending on what is used or choose the end property all together.

I mean: while there are different wishes, making it simpler now and better later is probably good. I am happy to support.

**Comment by tobixen** [comment]
The implementation details and the API should be decided upon.

It's not a lot of logic, but still ... there is some common logic that will apply both to `cal.Todo` and `cal.Event`.  Those inherits the `cal.Component` class.  The latter has 7 other child classes that don't have a DURATION (as far as I know ... don't have time to read up the RFC again on this one).

As for the implementation, I see some alternatives:

* Creating independent methods in `Todo` and `Event` classes.  There will be some duplicated logic.  It's not a lot of logic, but still ... I hate code duplication, so I don't like it :-)
* Creating public methods in the Component class ... which will raise an error if the component doesn't support DURATION.  I think it's not a good solution as we're "polluting" the namespace for many classes that don't need it.
* Creating public methods in `Event` and `Todo`, but putting common logic into "private" methods in the `Component` class.  I think I favor this approach.
* Make a new subclass between `Component` and `Todo`/`Event`.

**Comment by tobixen** [comment]
(came to think, I read some books about object-oriented design around 30 years ago, from that perspective the cleanest approach would be to have a separate subclass between `Component` and `Todo` / `Event` ...)

**Comment by tobixen** [comment]
Any opinions on the implementation details?

(I also touched upon this in a totally unrelated comment in the recurring-ical-event project, the component types VEVENT, VTODO and VJOURNAL is quite much related, can have almost all the same properties.  Both VTODO and VJOURNAL is "event-like", a task is basically an event with flexible start and end-time, and a journal is basically an event in the rearview mirror.  From that perspective a common superclass would be appropriate.  I should check the RFC to see if there is any naming convension there - otherwise, I would suggest something like "EventLike") ...

**Comment by niccokunzmann** [comment]
For me, superclass or shared function + polymorphy and other approaches are up to the implementing person/implementation detail. It would be nice to have a specification and tests.

I wonder about cases:
- start
- end
-start, end
-start, duration
-start,end,duration
-what is with period value type
-how does it relate to repetitions?

Would that be rather interesting for the return value of recurring-ical-events or also for general component with repetitions? Would it return a value or raise an error if there is unclarity?

**Comment by angatha** [comment]
I think introducing a more complex class hierarchy is not the way to go, if it does not yield benefits. Currently it takes not much effort to check the implementation for completenes aginast the RFC because the class definition of e.g. `Event` lists all properties.

For the case of duration/end getters and setters, I would add a fifth option:

Use pythons multiple inheritance feature to add the methods just to `cal.Todo` and `cal.Event` by making them inherit from a yet to be named class implementing the getters and setters:

```python
class Event(ClassImplementingMethids, Component):
    ...
```

**Comment by tobixen** [comment]
This one fell under the radar, sorry:

> It would be nice to have a specification and tests. 

It's often a nice practice to write the test code before the actual code ... but anyway, here some back-of-the-envelope specification/documentation/pseudocode - disregarding the Todo class as for now:

```python
class Event

    (...)

    @property
    def duration(self: Event) ->  datetime.timedelta
    """
    If 'DURATION' in self, return self['DURATION'].

    If 'DTSTART' and 'DTEND' in self, return self['DTEND'].dt-self['DTSTART'].dt

    Otherwise, return None
    """
    (...)

    @duration.setter
    def set_duration(self: Event, duration: datetime.timedelta, movable_attr: Literal['DTSTART', 'DTEND']='DTEND') -> None:
    """
    If movable_attr is DTEND and DTSTART is set (and for events, it should be set), set DTEND such that DTEND-'DTSTART == duration.

    If movable_attr is DTSTART and DTEND is set, set DTSTART such that self['DTEND']-self['DTSTART'] == duration.

    Otherwise, set self['DURATION'] = duration
    """

    @property
    def dtend(self: Event) -> datetime
    """"
    If 'DTEND' in self, return self['DTEND'].dt
    If 'DURATION' and 'DTSTART' in self, return DTSTART + DURATION
    Otherwise, return None
    """"
```

> I wonder about cases: - start - end -start, end -start, duration -start,end,duration -what is with period value type -how does it relate to repetitions?

The logic only touches DTSTART, DURATION and DTEND/DUE, so the results will be the same no matter if a RRULE is given or not.

**Comment by tobixen** [comment]
> Use pythons multiple inheritance feature to add the methods just to `cal.Todo` and `cal.Event` by making them inherit 

I'm not sure I feel comfortable with a MixIn-class.

Given that the code is quite straight-forward and trivial, and given that there will be some differences in the documentation (as well as the DTEND vs DUE difference), perhaps the best approach is simply to duplicate the code.  I hate code duplication, but it does make the code more readable in this case.

**Comment by niccokunzmann** [comment]
If you like, you can provide a pull request with the code that you feel comfortable with, using it.

I think, I would like to use VEVENT, VTODO and VJOURNAL with the same attributes, one for start, one for end and one for duration.

What could cause misunderstanding is if the event has RDATE as a period for example. Then, it is not clear which event you mean nor that the durations would match. If this is a feature that you would like to have for the recurring-ical-events library where the results do clearly have only one duration, start and end, then you could open an issue for that one.

**Comment by niccokunzmann** [comment]
> Otherwise, return None

I would opt for a timedelta of length zero.

Also, if the event DTSTART is on a date, the duration is one day. There are quite a few cases to test.

**Comment by tobixen** [comment]
Right, I should try to find some time for this.

I already wrote code in the caldav library, but it's not the right place for it, it's not very well designed, and it does need some more test code.  I'd gladly mark them as "deprecated" and let them call on methods in the icalendar library.

What I considered is to let `get_end`/`get_due` and `set_end`/`set_due` just be aliases, allowing the end user to do `event.set_due(...)` or `task.set_dtend(...)` and the library will just do the right thing.  I considered it to be an easy solution, a solution giving the best compatibility for end users (who don't need to worry if the object in the input parameter is a task or an event) and it also makes it possible to make nice code (the end user should not have to do `task.set_dtend(...)`)

**Comment by niccokunzmann** [comment]
Thinking of https://github.com/collective/icalendar/discussions/662:
I would like to use the `Event.dtstart` property.

Sanitizing input values and raising an error if they are invalid is a great feature.

also python-recurring-ical-events has calculations for start and end.
https://github.com/niccokunzmann/python-recurring-ical-events/blob/e91959369106031d61161014557817fbb3a94648/recurring_ical_events.py#L738

**Comment by niccokunzmann** [comment]
@tobixen, I started here: https://github.com/collective/icalendar/pull/721

**Comment by tobixen** [comment]
Thank you.  I've been quite swamped with other work and issues, so haven't had time to follow up.

For me the tasks are also important, with `DTSTART`, `DUE` and `DURATION`.

**Comment by niccokunzmann** [comment]
@tobixen 

> For me the tasks are also important, with DTSTART, DUE and DURATION.

I am working on this now.

How would you like the interface? I see that e.g. the VTODO as a
- DTSTART
- DUE
- COMPLETED

values

At the moment, I will do it so they are equivalent to the events: DTSTART == DTSTART, DTEND->DUE

However, I wonder what it is with the COMPLETED value. Could we assume something about this? E.g. what happens if the TODO was completed on another time? Could we assume that the end of the TODO is the COMPLETED value if given? It lasted that long, did it not?

**Comment by tobixen** [comment]
The relatively short answer, I'd like the interface as similar to the events as possible.  For algorithms to work that are agnostic on weather the object is an event or a task, I even think that it should deliver the due if one asks for  `.dtend`.

If changing duration, I think the DTSTART should be changed rather than DUE.  In my head, this is the key difference between events and tasks.  Events may come with a relatively important DTSTART, while the DTEND may be more of a guestimate. One may start with a task whenever one has available time, but sometimes the due timestamp is a hard due timestamp.  At the other hand, for the recurrence calculation it's the DTSTART that is important, always.

---

The long answer ... timestamps in tasks is a mess IMHO.

I'd like to record:

* A rough (or refined) time estimate (how long do I think it will take to do the task?)
* Timestamp for when I plan to start working on the task
* Timestamp for when I actually started working
* Timestamp for when I hope to be finished with the task
* Hard deadline for the task (and also: should the task be cancelled or procrastinated if the hard deadline was not met?)
* Timestamp for when I completed
* Actual time efficiently spent (and if this was professional work, it may be that the time we're billing for deviates from the work time actually spent on the task)

The RFC comes a bit short of offering all this, but the worst is that it gives very little information on what the attributes actually should be used for.  I've been wrangling my head around the attributes that are available and have come to think that this makes most sense:

* DTSTAMP - mandatory, created automatically and indicates when the ical entry was generated
* CREATED - non-mandatory.  If the calendar event has been copied from some old stone tablets, I suppose the timestamp should indicate when the event was actually created.
* LAST-MODIFIED - non-mandatory, I suppose it should be generated automatically.

Those three are relatively well-defined, but not helpful at all ref my wishlist above.

* DURATION - the only thing that is said in the RFC is that DURATION and DUE is mutually exclusive - so one may assume that difference between DUE and DTSTART should be equal to DURATION.  But what does DURATION mean?  Is it the estimated hours one needs to work actively with the task before it's completed?  Is it the number of days gone from one starts working with the task until it's expected to be completed?  I don't know - but the way I've considered it when developing `plann` is that the DURATION should be the estimated dedicated time needed to get the task done.
* DUE - as said, I think the DUE is a lot more important than DTSTART for tasks.  The DUE date may be a hard deadline, or it may be wishful thinking.  I'm using the PRIORITY field to indicate if it's a hard deadline or wishful thinking.
* DTSTART - if the DURATION is the difference between DUE and DTSTART, then DTSTART must be defined as the last possible moment one can theoretically sit down and start with the task to get it done in time.  Since humans needs to sleep as well, and since multiple tasks may have slightly overlapping DUE-time, the actual value of DTSTART then has no real practical meaning.
* COMPLETED could either be the automatically generated timestamp for when one clicks "I've done it!" in the calendaring tool, or it could be the actual timestamp for when the task was done.

I've considered that time tracking cannot and should not be stored in standard attributes in the task.  By definition, JOURNAL entries are about past events, unfortunately JOURNAL entries does not support DURATION or DTEND.

**Comment by niccokunzmann** [comment]
I split this up: 

journals, events and todos have a start, an end and a duration (lower case) rfc5545 properties don't calculate anything based on other properties but provide a typed interface. #662

**Comment by niccokunzmann** [comment]
@tobixen would you consider this closed? v6.0.2 is not released, yet.

**Comment by tobixen** [comment]
Sorry, this one fell through the cracks ... and ended up deep down in my todo-stack.  It looks nice, and I should try and replace "my" logic in the caldav library with calls to the icalendar library.

**Comment by tobixen** [comment]
```
    @property
    def start(self) -> date | datetime:
        """The start of the VTODO.
    
        Invalid values raise an InvalidCalendar.
        If there is no start, we also raise an IncompleteComponent error.
    
        You can get the start, end and duration of a Todo as follows:
    
        >>> from datetime import datetime
        >>> from icalendar import Todo
        >>> todo = Todo()
        >>> todo.start = datetime(2021, 1, 1, 12)
        >>> todo.end = datetime(2021, 1, 1, 12, 30) # 30 minutes
        >>> todo.duration  # 1800 seconds == 30 minutes
        datetime.timedelta(seconds=1800)
        >>> print(todo.to_ical())
        BEGIN:VTODO
        DTSTART:20210101T120000
        DUE:20210101T123000
        END:VTODO
        """
        start = self._get_start_end_duration()[0]
        if start is None:
>           raise IncompleteComponent("No DTSTART given.")
E           icalendar.error.IncompleteComponent: No DTSTART given.
```

I haven't studied the RFC5545 thoroughly, but in the caldav standard (which should be compatible with RFC5545) the duration is implicitly defined in https://datatracker.ietf.org/doc/html/rfc4791#section-9.9 ...

* Tasks have a zero duration if DUE is given but DTSTART is not given
* The duration of a task is defined to be the difference between CREATED and COMPLETED if those two are set
* If neither DUE nor DTSTART is set, then the task has an indefinite duration.

**Comment by tobixen** [comment]
Oups.  Wrong tag on the commit message.

**Comment by niccokunzmann** [comment]
@tobixen Could you check if we can close this or if anything needs doing?

**Comment by tobixen** [comment]
Checked from main 2e55d6e, 7.3.1.dev156
 
> * Tasks have a zero duration if DUE is given but DTSTART is not given

This does not seem to work:

```
>>> from icalendar import Todo
>>> from datetime import datetime,date
>>> todo=Todo()
>>> todo.start = datetime.now()
>>> todo.start
datetime.datetime(2026, 9, 6, 18, 56, 1, 807734)
>>> todo.duration
datetime.timedelta(0)
>>> todo=Todo()
>>> todo.end = datetime.now()
>>> todo.duration
Traceback (most recent call last):
  File "<python-input-10>", line 1, in <module>
    todo.duration
  File "/home/tobias/icalendar/src/icalendar/cal/todo.py", line 238, in duration
    return get_duration_property(self)
  File "/home/tobias/icalendar/src/icalendar/attr.py", line 1887, in get_duration_property
    return component.end - component.start
                           ^^^^^^^^^^^^^^^
  File "/home/tobias/icalendar/src/icalendar/cal/todo.py", line 202, in start
    return get_start_property(self)
  File "/home/tobias/icalendar/src/icalendar/attr.py", line 1822, in get_start_property
    raise IncompleteComponent(msg)
icalendar.error.IncompleteComponent: No DTSTART given.
```

The special case of due before start should result in zero duration, not negative duration if I remember correct.

```
>>> todo.start = datetime.now()
>>> todo.duration
datetime.timedelta(days=-1, seconds=86266, microseconds=907850)
```

> * Otherwise, the task has an indefinite duration.

I'm not sure if datetime.timedelta supports "infinite", and I'm not sure what is the best behaviour if so ... but possibly it should be handled in a better way than an exception.

```
>>> from icalendar import Todo
>>> todo=Todo()
>>> todo.duration
Traceback (most recent call last):
  File "<python-input-5>", line 1, in <module>
    todo.duration
  File "/home/tobias/icalendar/src/icalendar/cal/todo.py", line 238, in duration
    return get_duration_property(self)
  File "/home/tobias/icalendar/src/icalendar/attr.py", line 1887, in get_duration_property
    return component.end - component.start
           ^^^^^^^^^^^^^
  File "/home/tobias/icalendar/src/icalendar/cal/todo.py", line 216, in end
    return get_end_property(self, "DUE")
  File "/home/tobias/icalendar/src/icalendar/attr.py", line 1850, in get_end_property
    raise IncompleteComponent(msg)
icalendar.error.IncompleteComponent: No DUE or DURATION+DTSTART given.
>>> 
```

I'll ask the AI to generate some test code and throw it in as a draft pull request.

**Comment by tobixen** [comment]
I thought it was important that `task.duration` would not break down with an exception for tasks not having a `dtstart`-attribute - but maybe I was wrong.  A task with no dates set should probably still raise an exception - meaning that I anyway have to catch exceptions on the calling side, that makes this less important.

A draft pull request is made, with documentation on how I think it should work - but everything considered, I would not insist strongly on changing anything here.

**Comment by niccokunzmann** [comment]
Cool. Thanks for checking!

## Issue 1 (TODO)

Duration is 0 if only DUE is given. Correct?

```
>>> todo.end = datetime.now()
>>> todo.duration
timedelta(0)
```

## Issue 2 (AGREED)

duration must not be negative for TODO or any other component.
Usually, what I do is swap start and end of they are in the wrong order. What do you think?

## Issue 3 (TODO)

```
>>> todo=Todo()
>>> todo.duration
```

Ah. I remember:

>   A "VTODO" calendar component without the "DTSTART" and "DUE" (or
      "DURATION") properties specifies a to-do that will be associated
      with each successive calendar date, until it is completed.

So. What do you think? It could mean that it actually is only one day long. That or no date.


## Issue 4 - COMPLETED vs DUE (TODO)

>  COMPLETED:  This property defines the date and time that a to-do was
      actually completed.

So what is the duration of the TODO in this case?
How do DUE and COMPLETED influence the end, start and duration?

**Comment by tobixen** [comment]
> ## Issue 2 (AGREED)
> duration must not be negative for TODO or any other component. Usually, what I do is swap start and end of they are in the wrong order. What do you think?

I would be careful arbitrarily modifying data.  Currently icalendar can yield negative duration, and we decided it should be so in #999.  I have apparently overseen it, but RFC 5545 actually specifies that the DUE should be after DTSTART.

I've many times complained that the definition of DTSTART and DUE in a VTODO are vague.  It seems to make most sense (and most in accordance with the exact wording in the RFC) to use them for nailing the task as a planned activity on a calendar.  However, it's tempting to use DUE for a hard or soft deadline, or for a wish of the latest time a task should be completed - and except for very hard deadlines, it makes perfect sense to start with overdue tasks after the DUE-time.  But I digress.  Since it's already defined as negative, it makes sense leaving it there.

There exists a https://datatracker.ietf.org/doc/draft-ietf-calext-ical-tasks/ which may possibly improve the possibilities to store task data in icalendar format.

**Comment by tobixen** [comment]
> ## Issue 1 (TODO)
> Duration is 0 if only DUE is given. Correct?

Correct - and currently it raises an error.

> ## Issue 3 (TODO)
> ```
> >>> todo=Todo()
> >>> todo.duration
> ```
> 
> Ah. I remember:
> 
> > A "VTODO" calendar component without the "DTSTART" and "DUE" (or
> > "DURATION") properties specifies a to-do that will be associated
> > with each successive calendar date, until it is completed.
> 
> So. What do you think? It could mean that it actually is only one day long. That or no date.

"with each successive calendar date" could also mean "I should consider to work on this every day until it's completed".  I read it as "infinite duration", which cannot be expressed through a regular timedelta.  *Shrug* ... I think that raising an error probably makes more sense than guessing it's a full day task.  Or maybe not :-)

> ## Issue 4 - COMPLETED vs DUE (TODO)
> > COMPLETED:  This property defines the date and time that a to-do was
> > actually completed.
> 
> So what is the duration of the TODO in this case? How do DUE and COMPLETED influence the end, start and duration?

My instinct wants COMPLETED to have a higher priority than DUE, but as far as I read the RFCs, no.  "In a "VTODO" calendar component the property may be used to specify a duration for the to-do, instead of an explicit due DATE-TIME".  And then the search logic says that in the absence of duration and due, completed can be used. 

I think https://github.com/collective/icalendar/pull/1769/changes gives a relative good overview, it sort of works as a documentation on how I think it makes most sense to solve it.

**Comment by stevepiercy** [comment]
> I believe the intention is that an event with DTEND set one hour after the DTSTART is to be considered equivalent with an event with DURATION set to PT30M.

One hour == PT1H. @tobixen can you verify in the description?

> Issue 1

Agreed.

> Issue 2

I agree with @tobixen: don't swap the values. IMO, it's far more likely to be a typo or incorrectly selected input that wasn't validated by the client application, in which case the user would be grateful to receive feedback of their mistake.

> Issue 3

When `todo.duration` is undefined, I'd take two actions:

- [ ] Write to the RFC for clarification.
- [ ] Raise a [`UserWarning`](https://docs.python.org/3/library/exceptions.html#UserWarning) when `todo.duration` is undefined, perhaps `IcalendarDurationUndefined` with the message that https://datatracker.ietf.org/doc/html/rfc5545.html#section-3.6.2 is not clear. That way we can `¯\_(ツ)_/¯` without raising an exception or annoying the user, while complying with the RFC and passing the responsibility back upstream to resolve.

> Issue 4

https://github.com/collective/icalendar/pull/1769/changes is nice, but I'd prefer to refer to the VTODO table in https://datatracker.ietf.org/doc/html/rfc4791.html#section-9.9 for authoritative advice about how to handle a VTODO with and without DTSTART, DURATION, DUE, COMPLETED, or CREATED properties. If it's not in that table, then it's undefined.

Note that we might need to update the [documentation](https://icalendar.readthedocs.io/en/stable/reference/rfc-support.html) if we start to use RFC 4791 for guidance of an implementation.

**Comment by tobixen** [comment]
> > I believe the intention is that an event with DTEND set one hour after the DTSTART is to be considered equivalent with an event with DURATION set to PT30M.
> 
> One hour == PT1H. [@tobixen](https://github.com/tobixen) can you verify in the description?

Weird ... fixed.

**Comment by tobixen** [comment]
**I think we should just let things stay as they are.  It's simply not worth the effort nor discussion time - and whatever we do, it will be wrong**, or alternatively, I'd be happy to have a 0 duration returned when in doubt.

It would be nice if one could use this attribute without having to worry about exception-catching.  That's why I was negatively surprised that a `Todo` without DTSTART would raise an exception when asking for `task.duration`.

So a task without DUE, COMPLETED, DTSTART or START should be "associated with each successive calendar day", or appear on any date-related searches, which I interpret as "infinite duration".  From an engineering perspective it often makes sense to use numbers rather than mathematical concepts, and for all practical purposes "infinite" can often be rounded down to "a long timespan".  One option could be to calculate the time between DTSTAMP and now, or between some arbitrary historic timestamp, like the 1st of january 1970 or 2000 and now.  But this is also in the land of guesswork, which I think we should try to avoid.  Also, why should duration jump from "a very long timespan" to 0 just because someone slapped a due-date on the task?  It doesn't really make sense.  (it also doesn't make sense that a task disappears from the time-range-search of "things we ought to look into this week" just because someone slapped a DUE-date on it for next Monday, without also putting a DTSTART on the task, but that's what the caldav RFC says).

A side note: in my task-handling logic in `plann` I'm using the duration for storing time estimates - meaning that DTSTART is defined as "the latest possible time we can start working on the task, assuming there aren't any other overlapping events, tasks, chores, bedtime etc in the way" (a measure that doesn't make sense at all) - the new draft RFC comes with a new property for time estimates making this usage moot.  Anyway, with this usage of duration it would probably make more sense to catch an exception than to claim the estimated time for a task would be "infinite".

My negative surprise came from an attempt of replacing my own fetch-duration-logic with the icalendar logic and then my stuff blew up with the excpetion.  It seems that in my logic I just return 0 when the duration is not defined.  So from this perspective I'd be more happy to have 0 thrown in my face than an exception.

**Comment by tobixen** [comment]
Below is a table of different ways to interpret the timestamps (written by me, QA by Claude).

[`draft-ietf-calext-ical-tasks-17`](https://datatracker.ietf.org/doc/draft-ietf-calext-ical-tasks/) brings some clarity; the first line in the table below seems to be the indended usage of the timestamps, hence DURATION is the length of the "window" one has for getting the task done.  It also was my initial idea for plann, but the need for  storing time estimate somewhere won through and destroyed DTSTART.  However, the draft comes with a new property ESTIMATED-DURATION to be used for this.


| DTSTART | DUE | DURATION | PROBLEMS |
|---------|-----|----------|----------|
| The earliest time it makes sense to start on a task | The deadline | The "window" when one should consider working on the task | DTSTART may not be very well defined, DTSTART may need to be moved to an earlier time depending on how many other tasks one has on hand |
| The time one plans to start on the task | The deadline | The time between the planned start and the actual deadline.  It's not necessarily the time available for doing the task, as there may be other events, tasks, as well as things outside the calendar (e.g. bedtime) |  DURATION doesn't give much meaning |
| The time one actually started working on the task | The deadline | The time between the actual start and the deadline.  May be negative, when working on overdue tasks | Actual start may break recurring tasks |
| The last possible time one can start working on the task | The deadline | Time estimate including slack + time estimate for other competing tasks that also have to be done in the same period.  Not a very meaningful measurement. | DURATION doesn't give much meaning, and DTSTART ends up being a calculated value, which is even dependent on other tasks/events on the calendar. |
| The last possible time one can start working on the task, if this was the only thing on the calendar | The deadline | The time estimate for the task | DTSTART is destroyed, it's not a very meaningful measurement |
| The start of a time slot on the calendar intended for working on this task | The end of the time slot in which one intends to work on this task | Time estimate | The information about the deadline is lost, and one would quite often need to adjust the timestamps when the time ends up being spent on "more important things". |

**Comment by niccokunzmann** [comment]
@tobixen

I opened #1786 to track a subtask.
It makes sense for me to take on the changes that you propose because then they make sense to at least someone. I only use Events really. So, would you say that the tests in #1769 cover enough to close this issue? Then, we can say that they need to run for a PR and can open collaboration and fix the tests.

Also, as soon as you would like some functions for that RFC draft in icalendar, please open an issue. Can we expect that draft to go though or would there be an issue of having to remove code again in the future? Is the draft being used? It seems to be 11 years in the making with more recent uptake.

**Comment by tobixen** [comment]
As mentioned two steps above ... I don't think it's cost-effective to spend lots of efforts getting this "right".  At the end of the day, I think what really matters is ... should a consumer expect an exception if trying to access `task.duration`?

The easy way out, document that the consumer should be ready to handle an exception, keep logic as it is, drop the ball and close this issue (and the PR) right away, case closed.  

It's nice not to have to think of exceptions ... and it's nice that programs doesn't crash out of the blue just because some developer didn't care to catch exceptions.  In that case, we can return 0 instead of infinite for the "no dtstart, no due, nothing"-scenario.

**Comment by tobixen** [comment]
To summarize, as I see it this issue can be closed in three different ways - and it's needed with a decision from @niccokunzmann.  Thinking hard about it, my wish is the second option below:

- Close issue as-is, leave logic as it is, `task.duration` will raise an error for cases where the duration is not clearly defined.  Drawback: it may involve some pain on the receiving end as it's needed to handle exceptions (and there is also an asymmetry here if `event.duration` never raises an exception.  There is the risk that an application thoroughly tested with all kind of events will break down with an unhandled exception once someone throws in a simple task in the calendar).
- Return 0 in duration when the duration is not clearly defined.  This may not be entirely "correct", but it solves the "exception-pain" described above.
- Try to do things "as correct as possible", partly based on the interpretion of "duration" as the "window which the task can be performed", and conforming to how the CalDAV RFC specifies that search results should be returned, a task with no timestamps should be considered to have an infinite "window", and since "infinite" cannot be represented in a timedelta, again we need to raise an exception.

The third was my initial idea, but the more I think of it, the worse it seems.   The benefit is questionable - it would deliver a "correctness" that nobody (except me) asked for and that there exists no use-case for.  It does not solve the "exception pain" described in the first bullet point above, and it involves an implementation cost and added complexity in the library.

**Comment by niccokunzmann** [comment]
I think, practicality beats purity. Since you are for option 2 and that is practical based on experience, this solves it. For purists, we could add a little documentation on how to test for the special case. If they do not like it, they can open an issue. I think, this issue should be linked in the docs.
